macros/src/main/scala-3/com/softwaremill/macwire/internals/EligibleValuesFinder.scala

```diff
@@ -28,46 +28,67 @@ private[macwire] class EligibleValuesFinder[Q <: Quotes](log: Logger)(using val
-                )
-              case m: DefDef if m.termParamss.flatMap(_.params).isEmpty =>
-                EligibleValue(m.rhs.map(_.tpe).getOrElse(m.returnTpt.tpe), Select.unique(Ref(expr.symbol), m.name))
-            }
-            
-            Some(EligibleValues(Map(scope.widen -> r)))
+            val r = (expr.symbol.memberFields ::: expr.symbol.memberMethods)
+              .filter(m => notJunk(m) && isPublic(m))
+              .map(_.tree)
+              .collect {
+                case m: ValDef =>
+                  EligibleValue(
+                    m.tpt.tpe,
+                    This(expr.symbol.owner).select(expr.symbol).select(m.symbol)
+                  )
+                case m: DefDef if m.termParamss.flatMap(_.params).isEmpty =>
+                  EligibleValue(
+                    m.rhs.map(_.tpe).getOrElse(m.returnTpt.tpe),
+                    This(expr.symbol.owner).select(expr.symbol).select(m.symbol)
+                  )
+              }
+            EligibleValues(Map(scope.widen -> r))
           }
         } else {
-          None
+          EligibleValues.empty
         }
       }
 
       def buildEligibleValue(scope: Scope): PartialFunction[Tree, EligibleValues] = {
         case m: ValDef =>
           merge(
-            inspectModule(scope.widen, m.rhs.map(_.tpe).getOrElse(m.tpt.tpe), m).getOrElse(EligibleValues.empty),
-            EligibleValues(Map(scope -> List(EligibleValue(m.rhs.map(_.tpe).getOrElse(m.tpt.tpe), m))))
+            inspectModule(scope.widen, m.rhs.map(_.tpe).getOrElse(m.tpt.tpe), m),
+            EligibleValues(
+              Map(
+                scope -> List(
+                  EligibleValue(
+                    m.rhs.map(_.tpe).getOrElse(m.tpt.tpe),
+                    if symbol.isClassDef then This(symbol).select(m.symbol) else m
+                  )
+                )
+              )
+            )
           )
 
         case m: DefDef if m.termParamss.flatMap(_.params).isEmpty =>
           merge(
-            inspectModule(scope.widen, m.rhs.map(_.tpe).getOrElse(m.returnTpt.tpe), m).getOrElse(EligibleValues.empty),
-            EligibleValues(Map(scope -> List(EligibleValue(m.rhs.map(_.tpe).getOrElse(m.returnTpt.tpe), m))))
+            inspectModule(scope.widen, m.rhs.map(_.tpe).getOrElse(m.returnTpt.tpe), m),
+            EligibleValues(
+              Map(
+                scope -> List(
+                  EligibleValue(
+                    m.rhs.map(_.tpe).getOrElse(m.returnTpt.tpe),
+                    if symbol.isClassDef then This(symbol).select(m.symbol) else m
+                  )
+                )
+              )
+            )
           )
-
       }
 
       def handleClassDef(scope: Scope, s: Symbol): EligibleValues = {
-        (s.declaredMethods ::: s.declaredFields)
-          .filter(m => !m.fullName.startsWith("java.lang.Object") && !m.fullName.startsWith("scala.Any"))
+        (s.memberFields ::: s.memberMethods)
```

Comment:
I'm not sure this follows the rules as specified here: https://github.com/softwaremill/macwire#how-wiring-works (though I would expect to have tests for them); any values defined directly at the top-level should have precedence over inherited ones - at least that's how it was specified in Scala 2. So maybe we should first inspect the decalred methods/fields, and only later look at members?

Reply:
ok i'll try to add a test for that

Reply:
thanks! :)